File: web/src/sim/contentStore.ts
// Port of core/content_store.cpp: content-addressed storage. Writing a chunk
// whose hash already exists is a no-op, which gives deduplication for free.
// In the browser the "filesystem" is an in-memory map keyed by content-address.

export class ContentStore {
  private chunks = new Map<string, Uint8Array>();

  // True if a chunk with this hash is present.
  has(hash: string): boolean {
    return this.chunks.has(hash);
  }

  // Store a chunk. Returns true if newly written, false if the content-address
  // already existed (deduplicated).
  put(hash: string, data: Uint8Array): boolean {
    if (this.chunks.has(hash)) return false;
    this.chunks.set(hash, data.slice());
    return true;
  }

  // Read a chunk's bytes. Throws if missing (mirrors the C++ engine).
  get(hash: string): Uint8Array {
    const buf = this.chunks.get(hash);
    if (!buf) throw new Error("chunk not found in store: " + hash);
    return buf.slice();
  }

  // Number of unique chunks stored (distinct content-addresses).
  size(): number {
    return this.chunks.size;
  }

  // All stored content-addresses.
  hashes(): string[] {
    return [...this.chunks.keys()];
  }
}

File: core/src/content_store.cpp
#include "snapvault/content_store.h"

#include <stdexcept>

#include "snapvault/fs_util.h"

namespace snapvault {

ContentStore::ContentStore(std::string root)
    : root_(std::move(root)), chunks_dir_(fsutil::join(root_, "chunks")) {}

void ContentStore::init() const {
    fsutil::make_dirs(chunks_dir_);
}

std::string ContentStore::chunk_path(const std::string& hash) const {
    return fsutil::join(chunks_dir_, hash);
}

bool ContentStore::has(const std::string& hash) const {
    return fsutil::exists(chunk_path(hash));
}

bool ContentStore::put(const std::string& hash, const std::string& data) const {
    if (has(hash)) return false;  // dedup: identical content already stored
    fsutil::write_file(chunk_path(hash), data);
    return true;
}

std::string ContentStore::get(const std::string& hash) const {
    if (!has(hash)) {
        throw std::runtime_error("chunk not found in store: " + hash);
    }
    return fsutil::read_file(chunk_path(hash));
}

}  // namespace snapvault
